File: 0.1.1/activation_function/tanh.py
import numpy as np
from math import exp

def Function(input):
    return (exp(input)-exp(-input))/(exp(input)+exp(-input))

def Derivative(input):
    return 1-Function(input)**2

class tanh:

    def function(Q, start_position, batch_size, value_matrix):
        output_matrix = np.zeros(np.shape(value_matrix))
        for i in range(start_position, start_position+batch_size):
            for j in range(np.shape(output_matrix)[1]):
                output_matrix[i][j] = Function(value_matrix[i][j])
        Q.put(output_matrix)

    def derivative(inputs, position):
        derivative_vector = np.zeros(np.shape(inputs))
        derivative_vector[position] = Derivative(inputs[position])
        return derivative_vector

File: 0.2.4/activation_function/tanh.jl
using .Threads

module tanH
    function opt_diff(value)
        return 1-tanh(value)^2
    end

    function func(value_matrix::Array{Float32})
        output_matrix = zeros(Float32, size(value_matrix))
        Threads.@threads for i in eachindex(output_matrix)
            output_matrix[i] = tanh(value_matrix[i])
        end
        return output_matrix
    end

    function diff(inputs::Array{Float32}, position::Int64)
        derivative_vector = zeros(Float32, size(inputs))
        derivative_vector[position] = opt_diff(inputs[position])
        return derivative_vector
    end

    function get_name()
        return "tanH"
    end
end
